grub-core/loader/slaunch/x86_efi_linux.c

```diff
@@ -99,6 +99,16 @@ sl_efi_locate_mle_offset (struct grub_slaunch_params *slparams,
   slparams->mle_header_offset = mle_hdr_offset;
   slparams->mle_entry = mle_hdr->entry_point;
 
+  if (!is_linux) {
+    /*
+     * The previous value of the field is covering the whole EFI image which
+     * can include a lot of useless padding.  Limit the size used for measuring
+     * MLE to that reported by the header.  Don't change the behaviour for
+     * Linux.
+     */
+    slparams->mle_size = mle_hdr->mle_end - mle_hdr->mle_start;
```

Comment:
This is related to https://github.com/TrenchBoot/xen/pull/21#discussion_r1961703776. I think that field is used to check if PMR ranges are correct, so it should be made valid in the header.

Reply:
Interestingly, comparing `xen.efi.map` and `xen-syms.map`, it's not a 2 MiB difference but a bit over 3.5 MiB in my build.  However, it doesn't matter because linker works it out.  Sizes in headers of `xen` and `xen.efi` match symbol differences (`_end - start`) from corresponding map files.

Just to make it clear, entry points in headers don't differ just by 2 MiB, it's `0x428000` for my files.  `fixmlehdr` doesn't turn one header into another, linker resolves relocations as it should and yet `XEN_IMG_OFFSET` is not accounted for in `slaunch_stub_entry - start` in EFI case for some reason.  `xen.lds.S` and `efi.lds` have rather similar beginnings, so maybe linker just behaves differently in case of PE format.

Reply:
> Sizes in headers of xen and xen.efi match symbol differences (`_end - start`) from corresponding map files.

Yes, but shouldn't the size be `_end - (start - XEN_IMG_OFFSET)` under UEFI, to account for different base? Maybe I'm missing something, but I thought it looks like this currently:

```
                               size
                 /^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\
+----------------+----------------------------------+
|image base      |start                             |_end
+----------------+----------------------------------+
\__________________________________/
         measured part
```

where measured part size is short by that 2 MiB padding in the beginning.

Reply:
Actually, the field in MLE header is `Offset within MLE binary file`, so neither of those is valid, and that field shouldn't be used this way. As the file is loaded (not verbatim) by UEFI that doesn't understand MLE format, there is no use for this field at all, so it may stay as it is.